File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    bool solve(int i,int j,string &s,vector<vector<int>>&dp)
    {
        if(i>=j)
        {
            return 1;
        }
        
        if(dp[i][j]!=-1)return dp[i][j];
        
        
        if(s[i]==s[j])
        {
            bool x=solve(i+1,j-1,s,dp);
            
            dp[i][j]=x;
        }
        
        
         else{
             return dp[i][j]=false;
             
         }   
        
        return dp[i][j];
        
        
        
    }
    
    string longestPalindrome(string s) {
        
        vector<vector<int>>dp(1001,vector<int>(1001,-1));
        
        int maxi=0;
        
        int start=0;
        
        for(int i=0;i<s.length();i++)
        {
            for(int j=i;j<s.length();j++)
            {
                if(solve(i,j,s,dp))
                {
                    if(j-i+1>maxi)
                    {
                        maxi=j-i+1;
                        start=i;
                        
                    }
                }
            }
        }
        
        return s.substr(start,maxi);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    
    def solve(self,i,j,s,dp):
        
        if i>=j:
            return True
        
        
        
        if dp[i][j]!=-1:
            return dp[i][j]
        
        if s[i]==s[j]:
            
            dp[i][j] = self.solve(i + 1, j - 1, s, dp)
            
            
        
        else:
            
            dp[i][j]=False
            
            
        
        return dp[i][j]
    
    
    
    
    
    def longestPalindrome(self, s: str) -> str:
        
        n = len(s)
        
        dp = [[-1] * n for _ in range(n)]
        
        maxi = 0
        start = 0
        
        
        for i in range(n):
            for j in range(i,n):
                
                if self.solve(i,j,s,dp) and j-i+1>maxi:
                    
                    maxi=j-i+1
                    
                    start=i
    
       
        return s[start:start + maxi]
                
        
       